Synology DSM 7+ Fixes

johnnyxmas opened the pull request:
Add DSM 7+ Notifiarr install and uninstall scripts with log rotation and Entware support

- Installer auto-detects architecture (x86_64 / aarch64)
- Installs zstd via opkg (Entware) (Notifiarr no longer supplies tar.gz files for Synology architectures)
- Installs latest Notifiarr release from GitHub
- Redirects logs to /volume1/data to avoid disk quota issues
- Adds weekly log rotation (4 compressed backups)
- Creates DSM boot startup script and cron-based updater
- Includes uninstall script to fully remove all components -Updated README with new installation instructions

The old `install-synology.sh` script has been renames `install-synology-dsm6.sh`. NOTE: It is semi-broken (persistence does not work) as it relies on systemd functionality which DSM 6 (nor 7) does not have. This is NOT fixed in this commit.

johnnyxmas committed:
Synology DSM 7+ Fixes

Add DSM 7+ Notifiarr install and uninstall scripts with log rotation and Entware support

- Installer auto-detects architecture (x86_64 / aarch64)
- Installs zstd via opkg (Entware) (Notifiarr no longer supplies tar.gz files for Synology architectures)
- Installs latest Notifiarr release from GitHub
- Redirects logs to /volume1/data to avoid disk quota issues
- Adds weekly log rotation (4 compressed backups)
- Creates DSM boot startup script and cron-based updater
- Includes uninstall script to fully remove all components
-Updated README with new installation instructions

The old `install-synology.sh` script has been renames `install-synology-dsm6.sh`. NOTE: It is semi-broken (persistence does not work) as it relies on systemd functionality which DSM 6 (nor 7) does not have. This is NOT fixed in this commit.

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,155 @@
+
+echo "Extracting Notifiarr package..."
+if tar --help | grep -q -- '--zstd'; then
+  tar --zstd -xf "$PACKAGE"
```
Am I understanding correctly that you're downloading the Arch Linux installer package only to extract the binary from it? The `pkg.tar.zst` files are Arch Linux packages. If you need the binary, simply download the gz.

johnnyxmas replied:
Yep, per the commit message, there is no gz for x86_64 in the releases (aside from the raw source code), so we have to get the binary another way. I'd love to dispose of this workaround as it also forces the install of zstd which is a bad practice. Is there a  gz somewhere else (or am I blind)? I pull the list from[ here](https://github.com/Notifiarr/notifiarr/releases).

davidnewhall replied:
[This](https://github.com/Notifiarr/notifiarr/releases/download/v0.8.3/notifiarr.amd64.linux.gz) is the x86_64 file.

<img width="387" alt="Screenshot 2025-04-30 at 10 40 29 AM" src="https://github.com/user-attachments/assets/f0223dd8-901e-4ecf-8a9e-df9682377d28" />

johnnyxmas replied:
Ahhhhhhhh it's not expanding the list is the issue. Geez I spent way too long on that stupid workaround.

johnnyxmas replied:
I believe this is now fixed

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,155 @@
+
+CONFIGFILE=/etc/notifiarr/notifiarr.conf
+echo "Downloading clean default config from GitHub..."
+curl -sSLo "$CONFIGFILE" https://raw.githubusercontent.com/Notifiarr/notifiarr/main/examples/notifiarr.conf.example
```
Better would be to use the binary you just installed and `chmod` to be executable to generate this file. Find an example of that here: https://github.com/Notifiarr/notifiarr/pull/871/files

davidnewhall replied:
Passing `-w` will not start the service. It [writes the config](https://github.com/Notifiarr/notifiarr/blob/main/examples/MANUAL.md) file and exits.

johnnyxmas replied:
Done

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,155 @@
+
+echo "Creating config and log directories..."
+mkdir -p /etc/notifiarr /volume1/data
+chown -R root:root /etc/notifiarr /volume1/data
```
The config file needs to be writable by the user the app runs as.

johnnyxmas replied:
I believe I have fixed this

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,155 @@
+echo "10 3 * * * root /bin/bash $CRON_SCRIPT >> $CRON_LOG 2>&1" > "$CRON_FILE"
+
+echo "Installing weekly log rotation script..."
+LOGROTATE_SCRIPT="/usr/bin/rotate-notifiarr-logs.sh"
```
Notifiarr rotates its own log files. Recommend removing all this logic.

johnnyxmas replied:
Done

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,155 @@
+
+echo "Creating 'notifiarr' user if missing..."
+id notifiarr >/dev/null 2>&1 || {
+  PASS=$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 32)
```
It should not have a password and logins should be disabled.

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,155 @@
+cat << 'EOF' > /usr/local/etc/rc.d/notifiarr.sh
+#!/bin/sh
+case "$1" in
+  start)
```
This might be a good place to hard code the log file vars. Example in #871

johnnyxmas replied:
I believe I have corrected this.

davidnewhall replied:
I still suggest moving the `export` lines here.

johnnyxmas replied:
Moved them to notifiarr.sh

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,155 @@
+CRON_LOG="/volume1/data/notifiarr-updates.log"
+CRON_FILE="/etc/cron.d/update-notifiarr"
+
+curl -sSLo "$CRON_SCRIPT" https://raw.githubusercontent.com/Notifiarr/notifiarr/main/userscripts/install-synology.sh
```
The URL here is wrong.

johnnyxmas replied:
Fixed

davidnewhall reviewed (changes requested):
Thank you for this. Left a few comments. I do not have a syno and will need nitsua to test this.

johnnyxmas commented:
Thanks for this!! I agree this is all valid; just need clarification on the few comments I left above. 

FWIW I'm running DSM 7 and have done the install/uninstall several times as a test and things *appear* to be going well.

johnnyxmas committed:
Update install-synology-dsm7.sh

Added logic to handle updates seamlessly:
- Stops the existing notifiarr service before replacing the binary.
- Ensures the binary is replaced regardless of whether it already exists.
- Skips configuration file generation if it already exists.
Improved architecture detection:
- Ensures x86_64 uses the amd64.linux.gz binary.
Enhanced permissions and ownership:
- Ensures proper permissions for /usr/bin/notifiarr and /var/log/notifiarr.
- Sets stricter permissions for log directories.
Added robust error handling:
- Verifies the presence of required files and binaries after installation.
- Ensures the service is running after installation or update.
Updated cron job logic:
- Ensures the cron job script is downloaded and updated as needed.
Improved startup script:
- Ensures the service starts correctly after installation or update.
Removed unnecessary comments for cleaner code.

johnnyxmas committed:
Update install-synology-dsm7.sh

Made ARCH_NAME match binary naming conventions

Moved logs to a larger mount

Fixed user creation and permissions

Fixed update script URL

johnnyxmas committed:
Merge branch 'Notifiarr:main' into synology-dsm7-support

johnnyxmas committed:
Fix 'Text file busy' error in DSM7 installer by stopping service before extraction

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,137 @@
+chown -R notifiarr:notifiarr /volume1/@appdata/notifiarr/lognotifiarr /etc/notifiarr
+
+echo "Adding sudoers entry for smartctl..."
+sed -i '/notifiarr/d' /etc/sudoers
```
This will remove anything else with this word. I usually don't remove, and just use grep to check if it exists already before adding it with echo.

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,137 @@
+CRON_FILE="/etc/cron.d/update-notifiarr"
+
+curl -sSLo "$CRON_SCRIPT" https://raw.githubusercontent.com/Notifiarr/notifiarr/main/userscripts/install-synology-dsm7.sh
+chmod +x "$CRON_SCRIPT"
+echo "10 3 * * * root /bin/bash $CRON_SCRIPT >> /dev/null 2>&1" > "$CRON_FILE"
```
Bash has a $RANDOM variable you could make use of here so every syno notifiarr user doesn't hit GitHub at once.

johnnyxmas replied:
Hey, don't we all dream of a world where there would be enough notifiarr users to disrupt GitHub with a sync? 

I agree it's good to be polite, though. Fixed.

davidnewhall replied:
If this was _every_ client, they'd be pissed. Just syno? it's a pretty small foot print, but still nice to spread it out.

davidnewhall commented on userscripts/uninstall-notifiarr-dsm7.sh:
```diff
@@ -0,0 +1,23 @@
+rm -f /volume1/data/notifiarr-updates.log*
+
+echo "Removing Notifiarr user and sudoers entry..."
+sed -i '/notifiarr/d' /etc/sudoers
```
I would recommend grepping and printing the lines you remove before removing them.

johnnyxmas replied:
Fixed this, plus a number of inconsistencies between the install and uninstall scripts, as well as made the uninstaller a bit more robust.

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,137 @@
+set -e
+
+# Check for root privileges
+if [ "$EUID" -ne 0 ]; then
```
3-5ms in testing, but irrelevant in the scope of what this script does. EUID is fine.

johnnyxmas committed:
fix(synology): use grep instead of sed for sudoers entry in DSM7 installer

Replace sed command that removed all notifiarr entries with grep check that only adds the entry if missing. This prevents accidental removal of unrelated sudoers entries.

johnnyxmas committed:
Move log path env vars to startup script for persistence

johnnyxmas commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,137 @@
+  echo "Generating config file ${CONFIGFILE}"
+  echo " " > "${CONFIGFILE}"
+  export DN_LOG_FILE="/volume1/@appdata/notifiarr/lognotifiarr/app.log"
+  export DN_HTTP_LOG="/volume1/@appdata/notifiarr/lognotifiarr/http.log"
```
Really good call on that. Done.

johnnyxmas committed:
Randomize cron job timing to prevent simultaneous GitHub requests

davidnewhall commented:
I'll try to get @austinwbest to test this. He has a couple syncs, but is on holiday this week.

johnnyxmas committed:
Fix Synology DSM7 scripts: Add root check to uninstall, fix log paths, improve service stopping, make sudoers handling safer, add verification steps

davidnewhall commented on userscripts/uninstall-notifiarr-dsm7.sh:
```diff
@@ -0,0 +1,72 @@
+rm -f /usr/bin/notifiarr
+
+echo "Checking for /usr/bin/update-notifiarr.sh:"
+ls -la /usr/bin/update-notifiarr.sh 2>/dev/null || echo "File not found"
+rm -f /usr/bin/update-notifiarr.sh
```
```suggestion
[ -f "/usr/bin/update-notifiarr.sh" ]  || echo "File not found:  /usr/bin/update-notifiarr.sh"
```

davidnewhall commented on userscripts/uninstall-notifiarr-dsm7.sh:
```diff
@@ -0,0 +1,72 @@
+
+echo "Checking for /volume1/@appdata/notifiarr/lognotifiarr:"
+ls -la /volume1/@appdata/notifiarr/lognotifiarr 2>/dev/null || echo "Directory not found"
+rm -rf /volume1/@appdata/notifiarr/lognotifiarr
```
The `rm -rf` on line 42 does the same thing this one does, and more. Probably don't need both.

johnnyxmas replied:
Removed the earlier one.

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,150 @@
+  synouser --add notifiarr "" Notifiarr 0 "" 0
+fi
+
+echo "Setting permissions/ownership on: /usr/bin/notifiarr /volume1/@appdata/notifiarr/lognotifiarr"
```
There's not going to be a lot of files in `/volume1/@appdata/notifiarr`. I'd personally just put the logs in that base folder. Less typing to get to them.

johnnyxmas committed:
Switched to `-f` for file check

johnnyxmas committed:
Update userscripts/uninstall-notifiarr-dsm7.sh

Co-authored-by: David Newhall II <david@sleepers.pro>

johnnyxmas committed:
Merge branch 'synology-dsm7-support' of github.com:johnnyxmas/notifiarr into synology-dsm7-support

johnnyxmas committed:
Implemented `-f` and `-d` whwerever possible

johnnyxmas committed:
Eliminated redundant log directory check and removal

johnnyxmas committed:
ALways log to `/usr/bin/notifiarr`

johnnyxmas commented on userscripts/uninstall-notifiarr-dsm7.sh:
```diff
@@ -0,0 +1,72 @@
+
+echo "Removing Notifiarr files..."
+echo "Checking for /usr/bin/notifiarr:"
+ls -la /usr/bin/notifiarr 2>/dev/null || echo "File not found"
```
Learn and forget something every day. Changed.

johnnyxmas committed:
Added service restart capability and made readme more concise

johnnyxmas commented on README.md:
```diff
@@ -60,6 +60,47 @@ service notifiarr start
 
+## Synology DSM 6 (legacy)
+
+DSM 6 systems can install Notifiarr by downloading and extracting the appropriate `.tar.gz` binary from the [Releases](https://github.com/Notifiarr/notifiarr/releases/latest) page. You must set up a manual init script to make it persistent on reboot.
```
Upon further consideration, I think 90% of this info isn't something any user would need, and if they do, it's easy enough to search in the script. I've cleaned up the readme to be more concise.

davidnewhall commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,155 @@
+CRON_FILE="/etc/cron.d/update-notifiarr"
+
+# Download the update script - this should be a dedicated update script, not the installer
+curl -sSLo "$CRON_SCRIPT" https://raw.githubusercontent.com/Notifiarr/notifiarr/main/userscripts/update-notifiarr-dsm7.sh || \
```
I'm beginning to think there's a better way.

johnnyxmas replied:
Yeah, I caught this too. I'm currently working on it. I'm going prompt during install to have the user enable auto-update if desired. This will create a crontab to call `notifiarr.sh update`, which will cause the generated notifiarr.sh script to check if the latest release is newer, then download and install it. This also allows the user to still run `notifiarr.sh update` manually.

johnnyxmas replied:
That's the dream...and the right way to do this, but it's outside of my scope of knowledge. Here are the resources though: 

You have to make an SPK package (seems to be a proprietary docker container), then host it in a repo that users can add to their synos. [Synology has 2 pinned repos here](https://github.com/SynoCommunity) for each of those tasks. [This package development guide](https://help.synology.com/developer-guide/getting_started/first_package.html) is likely also useful.

johnnyxmas committed:
Update Function Fixes
- User is prompted to enable auto-update during install
- Update now runs via notfiarr.sh
- Improved upgrade process

Other:
- Removed obsolete syno scripts
- Renamed uninstaller to match installer naming convention

johnnyxmas committed:
keep only notifiarr.conf when performing preservation uninstall

johnnyxmas committed:
emoved duplicate stop message
Clearer verification output about config preservation
Added listing of preserved files when keeping configs
More accurate status messages during removal process

johnnyxmas committed:
- Use more flexible package URL matching patterns
- Add fallback URL matching logic
- Improve error messages with available download options
- Show GitHub API response when downloads fail

johnnyxmas committed:
Enhanced the installer script with:

Detailed debug output of GitHub API response
More robust asset parsing using jq
Case-insensitive package matching
Clearer error messages with available assets

johnnyxmas committed:
Fixed binary URL concatenation
Added option to manually provide binary

johnnyxmas committed:
Fixed download URL and checksum issues

johnnyxmas committed:
Remove file type check

johnnyxmas committed:
For the installer (install-synology-dsm7.sh):

Fixed the malformed URL problem by using direct URL construction with robust version extraction
Added multiple fallback mechanisms for version detection and checksum verification
Made the script more compatible with stock Synology DSM 7
Changed temporary file operations to use /etc/notifiarr/tmp instead of system /tmp
Fixed the error where the script was trying to create a directory at /usr/bin/notifiarr
Completely eliminated the crontab option since DSM doesn't have crontab:
Now only uses direct file creation in /etc/cron.d/ for automatic updates
For the uninstaller (uninstall-synology-dsm7.sh):

Fixed the issue with the binary not being deleted by using multiple removal attempts
Added a prompt to ask if the user wants to keep the configuration file:
If yes (default), preserves only notifiarr.conf and removes all other files
If no, removes the entire config directory including notifiarr.conf
Simplified the cron job cleanup to only check for and remove files in /etc/cron.d/
Added more robust process termination to ensure complete removal

johnnyxmas commented on userscripts/install-synology-dsm7.sh:
```diff
@@ -0,0 +1,155 @@
+
+# Set log file paths - these cannot be edited in the config
+export DN_LOG_FILE="/usr/bin/notifiarr/app.log"
+export DN_HTTP_LOG="/usr/bin/notifiarr/http.log"
```
Fixed

davidnewhall reviewed (changes requested):
Hey @johnnyxmas!

Austin and I finally jumped on a call today and discussed this contribution. We also helped a user troubleshoot their DSM 7 syno and get notifiarr running again. That user was able to run the existing script and it updated them correctly.

Our discussion lead to some questions about your contribution.

You mentioned a few things in your initial PR description that I need clarification on. One of the primary new features in DSM 7 is the replacement of upstart with systemd. You said DSM7 has no systemd, so I'm a bit confused on that. Which leads to further confusion because your pull request doesn't create a unit file for systemd. **What version of DSM are you running that lacks systemd?**

You also mentioned that the existing script is broken, but didn't provide errors or steps to reproduce the error. The existing scripts support DSM 6 and 7 by detecting the version and installing an upstart file or a systemd unit file. I've had no reports of problems and I just witnessed someone use it successfully. **Please let me know what problems you had with it.**

johnnyxmas commented:
Hey, apologies. The issue is more complex than "there is no `systemd`," and I may have even misspoke in saying that. I'll clear up the final ask first: **This all stemmed from me struggling with getting Notifiarr to maintain persistence across reboots or DSM updates.**  The issue had been going on since I was on 6, and persisted into 7, which is when I decided to start digging in. I, too, thought it was working fine with the existing script, except I noticed reboots and updates would often (always?) cause me to have to manually start the process. I got tired of it and began work on this script that would eliminate the problem.

Now, what I _should_ have said instead of "there is no `systemd`" is: DSM 7 _does not use_ `systemd`
While `systemctl` may exist on some DSM 7 devices, it's typically a stub or partial implementation. Furthermore, DSM 7 does not use `systemd` to manage services. Instead, Synology uses other control tools:

- BusyBox `init`
- Proprietary service tooling:
  - `/usr/syno/bin/synoservicectl`
  - `/usr/syno/bin/synopkg`
- Hardcoded startup hooks in `/etc/rc`, `/etc.defaults/rc`, and `/usr/syno/etc.defaults/rc.sysv/`


The following  are incomplete or ignored in DSM and risk being completely overwritten during DSM upgrades:
- `/etc/init.d/`exists but is not used by DSM 7
- `/etc/rc.local` may exist, but is ignored
- `systemd` unit files (`*.service`) 


For user-installed services, the only reliable (and supported) method is to place the startup scripts in `/usr/local/etc/rc.d/<your-script>.sh`. DSM7 executes all `*.sh` scripts in that dir on boot. 


Here's the relevant boot process, for clarity: 

1. Boot starts from BusyBox init
2. Calls `/etc/rc` scripts by runlevel (`rc.boot`, `rc.sysinit`, etc.)
3. When DSM boots, it eventually calls:
4. `/etc/rc` and includes `/usr/syno/sbin/synoservice`
5. That loads Synology-managed services
6. Then `/usr/local/etc/rc.d/*.sh` gets called
